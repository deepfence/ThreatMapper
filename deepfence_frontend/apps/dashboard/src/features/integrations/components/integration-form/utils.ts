import {
  ModelIntegrationFilters,
  ModelNodeIdentifierNodeTypeEnum,
} from '@/api/generated';
import { GenerativeAIIntegrationType, ScanTypeEnum } from '@/types/common';

export const IntegrationType = {
  slack: 'slack',
  pagerDuty: 'pagerduty',
  email: 'email',
  httpEndpoint: 'http_endpoint',
  microsoftTeams: 'teams',
  splunk: 'splunk',
  sumoLogic: 'sumologic',
  elasticsearch: 'elasticsearch',
  googleChronicle: 'googlechronicle',
  awsSecurityHub: 'aws_security_hub',
  jira: 'jira',
  s3: 's3',
} as const;
export type IntegrationKeyType = (typeof IntegrationType)[keyof typeof IntegrationType];
export function getIntegrationPrettyName(type: IntegrationKeyType) {
  switch (type) {
    case IntegrationType.slack:
      return 'Slack';
    case IntegrationType.pagerDuty:
      return 'PagerDuty';
    case IntegrationType.httpEndpoint:
      return 'HTTP Endpoint';
    case IntegrationType.microsoftTeams:
      return 'Microsoft Teams';
    case IntegrationType.splunk:
      return 'Splunk';
    case IntegrationType.sumoLogic:
      return 'Sumo Logic';
    case IntegrationType.elasticsearch:
      return 'Elasticsearch';
    case IntegrationType.googleChronicle:
      return 'Google Chronicle';
    case IntegrationType.awsSecurityHub:
      return 'AWS Security Hub';
    case IntegrationType.jira:
      return 'Jira';
    case IntegrationType.s3:
      return 'S3';
    case IntegrationType.email:
      return 'Email';
    default:
      throw new Error('Integration type not found');
  }
}

export const AIIntegrationDocsLinkMap: Record<GenerativeAIIntegrationType, string> = {
  openai: 'https://community.deepfence.io/threatmapper/docs/integrations/threatrx',
  'amazon-bedrock':
    'https://community.deepfence.io/threatmapper/docs/integrations/threatrx',
};

export const IntegrationDocsLinkMap: Record<IntegrationKeyType, string> = {
  slack: 'https://community.deepfence.io/threatmapper/docs/integrations/slack',
  pagerduty: 'https://community.deepfence.io/threatmapper/docs/integrations/pagerduty',
  email: 'https://community.deepfence.io/threatmapper/docs/integrations/email',
  http_endpoint:
    'https://community.deepfence.io/threatmapper/docs/integrations/http-endpoint',
  teams: 'https://community.deepfence.io/threatmapper/docs/integrations/microsoft-teams',
  splunk: 'https://community.deepfence.io/threatmapper/docs/integrations/splunk',
  sumologic: 'https://community.deepfence.io/threatmapper/docs/integrations/sumo-logic',
  elasticsearch:
    'https://community.deepfence.io/threatmapper/docs/integrations/elasticsearch',
  googlechronicle: '',
  aws_security_hub: '',
  jira: 'https://community.deepfence.io/threatmapper/docs/integrations/jira',
  s3: 'https://community.deepfence.io/threatmapper/docs/integrations/s3',
} as const;

export const isCloudTrailNotification = (notificationType: string) => {
  return notificationType && notificationType === 'CloudTrailAlert';
};

export const isUserActivityNotification = (notificationType: string) => {
  return notificationType && notificationType === 'UserActivities';
};

export const isVulnerabilityNotification = (notificationType: string) => {
  return notificationType && notificationType === 'Vulnerability';
};

export const isJiraIntegration = (integrationType: string) => {
  return integrationType && integrationType === IntegrationType.jira;
};

export const isCloudComplianceNotification = (notificationType: string) => {
  return (
    notificationType &&
    (notificationType.toLowerCase() === 'cloudcompliance' ||
      notificationType.toLowerCase() === 'cloud_compliance')
  );
};

export const isComplianceNotification = (notificationType: string) => {
  return notificationType && notificationType === 'Compliance';
};

export const getHostsFilter = (nodeIds: ModelIntegrationFilters['node_ids'] = []) => {
  if (!nodeIds) {
    return [];
  }
  return nodeIds.reduce((acc: string[], current) => {
    if (current.node_type === ModelNodeIdentifierNodeTypeEnum.Host) {
      acc.push(current.node_id);
    }
    return acc;
  }, []);
};

export const getImagesFilter = (nodeIds: ModelIntegrationFilters['node_ids'] = []) => {
  if (!nodeIds) {
    return [];
  }
  return nodeIds.reduce((acc: string[], current) => {
    if (current.node_type === ModelNodeIdentifierNodeTypeEnum.Image) {
      acc.push(current.node_id);
    }
    return acc;
  }, []);
};

export const getClustersFilter = (nodeIds: ModelIntegrationFilters['node_ids'] = []) => {
  if (!nodeIds) {
    return [];
  }
  return nodeIds.reduce((acc: string[], current) => {
    if (current.node_type === ModelNodeIdentifierNodeTypeEnum.Cluster) {
      acc.push(current.node_id);
    }
    return acc;
  }, []);
};

export const getCloudAccountsFilter = (
  nodeIds: ModelIntegrationFilters['node_ids'] = [],
) => {
  if (!nodeIds) {
    return [];
  }
  return nodeIds.reduce((acc: string[], current) => {
    if (current.node_type === ModelNodeIdentifierNodeTypeEnum.CloudAccount) {
      acc.push(current.node_id);
    }
    return acc;
  }, []);
};

export const API_SCAN_TYPE_MAP: Record<string, ScanTypeEnum> = {
  Vulnerability: ScanTypeEnum.VulnerabilityScan,
  Secret: ScanTypeEnum.SecretScan,
  Malware: ScanTypeEnum.MalwareScan,
  Compliance: ScanTypeEnum.ComplianceScan,
};

export const scanTypes = ['Secret', 'Vulnerability', 'Malware'];

export const getNotificationPrettyName = (notificationType: string) => {
  if (isCloudTrailNotification(notificationType)) {
    return 'CloudTrail Alert';
  } else if (isUserActivityNotification(notificationType)) {
    return 'User Activities';
  } else if (isCloudComplianceNotification(notificationType)) {
    return 'Cloud Compliance';
  }
  return notificationType;
};

export const canSendScanSummary = (notificationType: string, integrationType: string) => {
  return (
    (integrationType === IntegrationType.slack ||
      integrationType === IntegrationType.microsoftTeams) &&
    (['Secret', 'Vulnerability', 'Malware'].includes(notificationType) ||
      isComplianceNotification(notificationType) ||
      isCloudComplianceNotification(notificationType))
  );
};
