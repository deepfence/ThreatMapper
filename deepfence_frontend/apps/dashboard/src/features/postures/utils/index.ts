import {
  ModelCloudComplianceStatusEnum,
  ModelCloudNodeAccountsListReqCloudProviderEnum,
  ModelComplianceStatusEnum,
  ModelScanInfoStatusEnum,
} from '@/api/generated';

export const isCloudNonOrgNode = (nodeType?: string) => {
  return (
    nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Aws ||
    nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Azure ||
    nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Gcp
  );
};

export const isCloudOrgNode = (nodeType?: string) =>
  nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.AwsOrg ||
  nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.GcpOrg ||
  nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.AzureOrg;

export const isNonCloudNode = (nodeType: string) => {
  return (
    nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Linux ||
    nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Kubernetes
  );
};

export const isLinuxNodeType = (nodeType: string) =>
  nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Linux;

export const isKubernetesNodeType = (nodeType: string) =>
  nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Kubernetes;

export const isAlarmStatus = (status: string) => {
  return status?.toLowerCase() === ModelCloudComplianceStatusEnum.Alarm;
};
export const isInfoStatus = (status: string) => {
  return status?.toLowerCase() === ModelCloudComplianceStatusEnum.Info;
};
export const isOkStatus = (status: string) => {
  return status?.toLowerCase() === ModelCloudComplianceStatusEnum.Ok;
};
export const isSkipStatus = (status: string) => {
  return status?.toLowerCase() === ModelCloudComplianceStatusEnum.Skip;
};
export const isPassStatus = (status: string) => {
  return status?.toLowerCase() === ModelComplianceStatusEnum.Pass;
};
export const isWarnStatus = (status: string) => {
  return status?.toLowerCase() === ModelComplianceStatusEnum.Warn;
};
export const isNoteStatus = (status: string) => {
  return status?.toLowerCase() === ModelComplianceStatusEnum.Note;
};
export const isDeleteStatus = (status: string) => {
  return (
    status?.toLowerCase() === ModelCloudComplianceStatusEnum.Delete ||
    status?.toLowerCase() === ''
  );
};

export function getDisplayNameOfNodeType(
  nodeType?: ModelCloudNodeAccountsListReqCloudProviderEnum,
) {
  if (!nodeType) {
    throw new Error(`Node type cannot be empty for display name`);
  }

  if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Azure) {
    return 'Subscription';
  } else if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.AzureOrg) {
    return 'Tenant';
  } else if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Gcp) {
    return 'Project';
  } else if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.GcpOrg) {
    return 'Organization Project';
  } else {
    return 'Account';
  }
}

export function getSearchableCloudAccountDisplayName(
  nodeType?: ModelCloudNodeAccountsListReqCloudProviderEnum,
) {
  if (!nodeType) {
    throw new Error(`Node type cannot be empty for display name`);
  }
  if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Azure) {
    return 'Subscription';
  } else if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.AzureOrg) {
    return 'Tenant';
  } else if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Gcp) {
    return 'Project';
  } else if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.GcpOrg) {
    return 'Organization project';
  } else if (isCloudOrgNode(nodeType)) {
    return 'Organization account';
  } else {
    return 'Account';
  }
}

export function getDeleteConfirmationDisplayName(
  nodeType?: ModelCloudNodeAccountsListReqCloudProviderEnum,
) {
  if (!nodeType) {
    throw new Error(`Node type cannot be empty for display name`);
  }
  if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Azure) {
    return 'The Selected subscription, resources and scans related to the subscription will be deleted.';
  } else if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.AzureOrg) {
    return 'The Selected tenant, child subscriptions related to tenant, resources and scans related to tenant will be deleted.';
  } else if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.Gcp) {
    return 'The Selected project, resources and scans related to the project will be deleted.';
  } else if (nodeType === ModelCloudNodeAccountsListReqCloudProviderEnum.GcpOrg) {
    return 'The Selected organization project, child projects related to organization project, resources and scans related to organization project will be deleted.';
  } else if (isCloudOrgNode()) {
    return 'The Selected organization cloud account, child accounts related to organization account, resources and scans related to the cloud accounts will be deleted.';
  } else {
    return 'The Selected cloud account, resources and scans related to the account will be deleted.';
  }
}

export const isRefreshAccountFailed = (status: string): boolean => {
  if (status?.length && ModelScanInfoStatusEnum.Error === status) {
    return true;
  }
  return false;
};
