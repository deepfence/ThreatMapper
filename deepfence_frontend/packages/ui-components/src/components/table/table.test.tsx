import '@testing-library/jest-dom';

import { RowSelectionState, SortingState } from '@tanstack/react-table';
import { waitFor } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { sortBy } from 'lodash-es';
import { useMemo, useState } from 'react';

import {
  getRowExpanderColumn,
  getRowSelectionColumn,
  Table,
  TableProps,
} from '@/components/table/Table';
import { renderUI } from '@/tests/utils';

interface Fruit {
  id: number;
  name: string;
  description: string;
}

function createDummyFruitData(length = 2) {
  const data: Fruit[] = [];
  for (let i = 0; i < length; i++) {
    data.push({
      id: i,
      name: `Fruit ${i}`,
      description: `Description for ${i}`,
    });
  }
  return data;
}

const defaultColumns = [
  {
    accessorKey: 'id',
  },
  {
    accessorKey: 'name',
  },
  {
    accessorKey: 'description',
    enableResizing: false,
  },
];

const BasicTable = ({
  dataLength,
  tableProps,
}: {
  dataLength?: number;
  tableProps?: Partial<TableProps<Fruit>>;
}) => {
  return (
    <Table
      data={createDummyFruitData(dataLength)}
      columns={defaultColumns}
      {...tableProps}
    />
  );
};

describe(`Component Table`, () => {
  it('should render a basic table', async () => {
    const user = userEvent.setup();
    const { getByRole, getAllByTestId, getAllByRole } = renderUI(<BasicTable />);
    expect(getByRole('table')).toBeInTheDocument();
    expect(getAllByTestId('table-header-row').length).toEqual(1);
    expect(getAllByRole('columnheader').length).toEqual(3);
    const rows = getAllByRole('row');
    expect(rows.length).toEqual(3);
    const cells = getAllByRole('cell');
    expect(cells.length).toEqual(6);
    cells.forEach((cell, index) => {
      if (index < 3) {
        expect(cell.outerHTML).toMatchSnapshot(`cells with border ${index}`);
      } else {
        expect(cell.outerHTML).toMatchSnapshot(`cells without border ${index}`);
      }
    });
    await user.hover(rows[2]);

    await waitFor(() => {
      expect(rows[2]).toMatchSnapshot('row with hover state on normal table');
    });
  });

  it('expandable rows should expand correctly', async () => {
    const user = userEvent.setup();
    const { getAllByRole, getByText, queryByText } = renderUI(
      <BasicTable
        dataLength={5}
        tableProps={{
          renderSubComponent: ({ row }) => {
            return <div>row id {row.original.id}</div>;
          },
          getRowCanExpand: (row) => {
            return row.index !== 0;
          },
          columns: [getRowExpanderColumn<Fruit>(), ...defaultColumns],
        }}
      />,
    );
    const expanderButtons = getAllByRole('button');
    expect(expanderButtons.length).toEqual(4);

    expect(queryByText('row id 2')).not.toBeInTheDocument();

    user.click(expanderButtons[1]);

    await waitFor(() => {
      expect(getByText('row id 2')).toBeInTheDocument();
    });
  });

  it('table with auto pagination should work', async () => {
    const user = userEvent.setup();
    const { getByTestId, getByRole, queryByText, rerender } = renderUI(
      <BasicTable
        dataLength={100}
        tableProps={{
          enablePagination: true,
          pageSize: 20,
        }}
      />,
    );

    expect(getByTestId('pagination-container')).toBeInTheDocument();
    expect(getByRole('button', { name: /1/i })).toBeInTheDocument();
    expect(getByRole('button', { name: /5/i })).toBeInTheDocument();
    expect(getByTestId('pagination-next')).toBeInTheDocument();
    expect(getByTestId('pagination-prev')).toBeInTheDocument();
    expect(getByRole('cell', { name: /fruit 0/i })).toBeInTheDocument();

    user.click(getByRole('button', { name: /2/i }));

    await waitFor(() => {
      expect(queryByText('fruit 0')).not.toBeInTheDocument();
      expect(getByRole('cell', { name: /fruit 20/i })).toBeInTheDocument();
    });

    rerender(
      <BasicTable
        dataLength={100}
        tableProps={{
          enablePagination: true,
          pageSize: 10,
        }}
      />,
    );

    await waitFor(() => {
      expect(getByRole('button', { name: /10/i })).toBeInTheDocument();
    });
  });

  it('table with manual pagination should work', async () => {
    const user = userEvent.setup();

    const ManualPaginationTable = () => {
      const [{ pageIndex, pageSize }, setPagination] = useState({
        pageIndex: 0,
        pageSize: 20,
      });
      const data = useMemo(() => {
        const data = createDummyFruitData(995);
        return data.slice(pageIndex * pageSize, (pageIndex + 1) * pageSize);
      }, [pageIndex]);

      return (
        <Table
          data={data}
          columns={defaultColumns}
          enablePagination
          manualPagination
          pageSize={pageSize}
          pageIndex={pageIndex}
          totalRows={995}
          onPaginationChange={setPagination}
        />
      );
    };

    const { getByTestId, getByRole, queryByText } = renderUI(<ManualPaginationTable />);

    expect(getByTestId('pagination-container')).toBeInTheDocument();
    expect(getByRole('button', { name: /1/i })).toBeInTheDocument();
    expect(getByRole('button', { name: /50/i })).toBeInTheDocument();
    expect(getByTestId('pagination-next')).toBeInTheDocument();
    expect(getByTestId('pagination-prev')).toBeInTheDocument();
    expect(getByRole('cell', { name: /fruit 0/i })).toBeInTheDocument();

    user.click(getByRole('button', { name: /2/i }));

    await waitFor(() => {
      expect(queryByText('fruit 0')).not.toBeInTheDocument();
      expect(getByRole('cell', { name: /fruit 20/i })).toBeInTheDocument();
    });
  });

  it('resizable table should show resize handlers on headers', () => {
    const { getByTestId, queryByTestId } = renderUI(
      <BasicTable
        dataLength={10}
        tableProps={{
          enableColumnResizing: true,
        }}
      />,
    );
    expect(getByTestId('column-resizer-id')).toBeInTheDocument();
    expect(getByTestId('column-resizer-name')).toBeInTheDocument();
    expect(queryByTestId('column-resizer-description')).not.toBeInTheDocument();
  });

  it('table with automatic sorting should work correctly', async () => {
    const user = userEvent.setup();
    const { getAllByRole, getByRole, getByTestId, queryByTestId } = renderUI(
      <BasicTable
        dataLength={100}
        tableProps={{
          enableSorting: true,
          enablePagination: true,
          pageSize: 10,
        }}
      />,
    );
    const rows = getAllByRole('row');
    expect(rows.length).toEqual(11);
    expect(rows[1].children.item(0)?.textContent).toEqual('0');
    expect(rows[10].children.item(0)?.textContent).toEqual('9');
    expect(getByTestId('column-unsorted-indicator-id')).toBeInTheDocument();
    user.click(getByRole('columnheader', { name: /id/i }));

    await waitFor(() => {
      const rows = getAllByRole('row');
      expect(queryByTestId('column-unsorted-indicator-id')).not.toBeInTheDocument();
      expect(queryByTestId('column-ascending-indicator-id')).not.toBeInTheDocument();
      expect(getByTestId('column-descending-indicator-id')).toBeInTheDocument();
      expect(rows.length).toEqual(11);
      expect(rows[1].children.item(0)?.textContent).toEqual('99');
      expect(rows[10].children.item(0)?.textContent).toEqual('90');
    });

    user.click(getByRole('columnheader', { name: /id/i }));

    await waitFor(() => {
      const rows = getAllByRole('row');
      expect(queryByTestId('column-unsorted-indicator-id')).not.toBeInTheDocument();
      expect(queryByTestId('column-descending-indicator-id')).not.toBeInTheDocument();
      expect(getByTestId('column-ascending-indicator-id')).toBeInTheDocument();
      expect(rows.length).toEqual(11);
      expect(rows[1].children.item(0)?.textContent).toEqual('0');
      expect(rows[10].children.item(0)?.textContent).toEqual('9');
    });

    user.click(getByRole('columnheader', { name: /id/i }));

    await waitFor(() => {
      const rows = getAllByRole('row');
      expect(queryByTestId('column-ascending-indicator-id')).not.toBeInTheDocument();
      expect(queryByTestId('column-descending-indicator-id')).not.toBeInTheDocument();
      expect(getByTestId('column-unsorted-indicator-id')).toBeInTheDocument();
      expect(rows.length).toEqual(11);
      expect(rows[1].children.item(0)?.textContent).toEqual('0');
      expect(rows[10].children.item(0)?.textContent).toEqual('9');
    });
  });

  it('table with manual sorting should work correctly', async () => {
    const user = userEvent.setup();
    const ManualSortedTable = () => {
      const [{ pageIndex, pageSize }, setPagination] = useState({
        pageIndex: 0,
        pageSize: 10,
      });
      const [sort, setSort] = useState<SortingState>([]);
      const data = useMemo(() => {
        let data = createDummyFruitData(100);
        if (sort.length) {
          data = sortBy(data, [sort[0].id]);
          if (sort[0].desc) {
            data.reverse();
          }
        }
        return data.slice(pageIndex * pageSize, (pageIndex + 1) * pageSize);
      }, [pageIndex, sort]);

      return (
        <Table
          data={data}
          columns={defaultColumns}
          enablePagination
          manualPagination
          pageSize={pageSize}
          pageIndex={pageIndex}
          totalRows={100}
          onPaginationChange={setPagination}
          enableSorting
          manualSorting
          sortingState={sort}
          onSortingChange={setSort}
        />
      );
    };
    const { getAllByRole, getByRole, getByTestId, queryByTestId } = renderUI(
      <ManualSortedTable />,
    );
    const rows = getAllByRole('row');
    expect(rows.length).toEqual(11);
    expect(rows[1].children.item(0)?.textContent).toEqual('0');
    expect(rows[10].children.item(0)?.textContent).toEqual('9');
    expect(getByTestId('column-unsorted-indicator-id')).toBeInTheDocument();
    user.click(getByRole('columnheader', { name: /id/i }));

    await waitFor(() => {
      const rows = getAllByRole('row');
      expect(queryByTestId('column-unsorted-indicator-id')).not.toBeInTheDocument();
      expect(queryByTestId('column-ascending-indicator-id')).not.toBeInTheDocument();
      expect(getByTestId('column-descending-indicator-id')).toBeInTheDocument();
      expect(rows.length).toEqual(11);
      expect(rows[1].children.item(0)?.textContent).toEqual('99');
      expect(rows[10].children.item(0)?.textContent).toEqual('90');
    });

    user.click(getByRole('columnheader', { name: /id/i }));

    await waitFor(() => {
      const rows = getAllByRole('row');
      expect(queryByTestId('column-unsorted-indicator-id')).not.toBeInTheDocument();
      expect(queryByTestId('column-descending-indicator-id')).not.toBeInTheDocument();
      expect(getByTestId('column-ascending-indicator-id')).toBeInTheDocument();
      expect(rows.length).toEqual(11);
      expect(rows[1].children.item(0)?.textContent).toEqual('0');
      expect(rows[10].children.item(0)?.textContent).toEqual('9');
    });

    user.click(getByRole('columnheader', { name: /id/i }));

    await waitFor(() => {
      const rows = getAllByRole('row');
      expect(queryByTestId('column-ascending-indicator-id')).not.toBeInTheDocument();
      expect(queryByTestId('column-descending-indicator-id')).not.toBeInTheDocument();
      expect(getByTestId('column-unsorted-indicator-id')).toBeInTheDocument();
      expect(rows.length).toEqual(11);
      expect(rows[1].children.item(0)?.textContent).toEqual('0');
      expect(rows[10].children.item(0)?.textContent).toEqual('9');
    });
  });

  it('table with row selection enabled should work correctly', async () => {
    const user = userEvent.setup();
    const TableWithRowSelection = () => {
      const [rowSelectionState, setRowSelectionState] = useState<RowSelectionState>({});
      return (
        <>
          <div data-testid="selected-rows">
            {Object.keys(rowSelectionState)
              .map((id) => `"${id}"`)
              .join(', ')}
          </div>
          <BasicTable
            dataLength={10}
            tableProps={{
              columns: [getRowSelectionColumn(), ...defaultColumns],
              enableSorting: true,
              enablePagination: true,
              pageSize: 10,
              enableRowSelection: true,
              rowSelectionState,
              onRowSelectionChange: setRowSelectionState,
              getRowId: ({ id }) => {
                return `id-${id}`;
              },
            }}
          />
        </>
      );
    };
    const { getAllByRole, getByTestId } = renderUI(<TableWithRowSelection />);

    let [selectAllCheckBox, ...checkboxes] = getAllByRole('checkbox');
    expect(selectAllCheckBox).toBeInTheDocument();
    expect(checkboxes.length).toEqual(10);
    expect(getByTestId('selected-rows')).toHaveTextContent('');

    user.click(checkboxes[4]);

    await waitFor(() => {
      expect(getByTestId('selected-rows').textContent).toContain(`"id-4"`);
    });

    [selectAllCheckBox, ...checkboxes] = getAllByRole('checkbox');
    user.click(checkboxes[5]);

    await waitFor(() => {
      expect(getByTestId('selected-rows').textContent).toContain(`"id-4"`);
      expect(getByTestId('selected-rows').textContent).toContain(`"id-5"`);
    });

    [selectAllCheckBox, ...checkboxes] = getAllByRole('checkbox');
    user.click(selectAllCheckBox);

    await waitFor(() => {
      checkboxes.slice(checkboxes.length - 1).forEach((checkbox, index) => {
        expect(getByTestId('selected-rows').textContent).toContain(`"id-${index}"`);
      });
    });
  });

  it('should use getTrProps and getTdProps correctly', async () => {
    const { getByRole, getAllByTestId, getAllByRole } = renderUI(
      <BasicTable
        tableProps={{
          getTdProps: () => {
            return {
              className: 'text-red-500',
            };
          },
          getTrProps: () => {
            return {
              className: 'bg-red-500',
            };
          },
        }}
      />,
    );
    expect(getByRole('table')).toBeInTheDocument();
    expect(getAllByTestId('table-header-row').length).toEqual(1);
    expect(getAllByRole('columnheader').length).toEqual(3);
    const rows = getAllByRole('row');
    expect(rows.length).toEqual(3);
    const cells = getAllByRole('cell');
    expect(cells.length).toEqual(6);
    rows.forEach((row, index) => {
      expect(row.outerHTML).toMatchSnapshot(`row with custom props${index}`);
    });
    cells.forEach((cell, index) => {
      expect(cell.outerHTML).toMatchSnapshot(`cells with custom props ${index}`);
    });
  });
});
