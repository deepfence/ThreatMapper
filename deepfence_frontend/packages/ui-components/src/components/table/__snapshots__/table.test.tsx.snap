// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Component Table > should render a basic table > cells with border 0 1`] = `"<td style=\\"width: 150px;\\" class=\\"text-[13px] font-medium leading-[18px] text-text-text-and-icon px-4 truncate min-w-0 border-b border-bg-grid-border h-[48px]\\">0</td>"`;

exports[`Component Table > should render a basic table > cells with border 1 1`] = `"<td style=\\"width: 150px;\\" class=\\"text-[13px] font-medium leading-[18px] text-text-text-and-icon px-4 truncate min-w-0 border-b border-bg-grid-border h-[48px]\\">Fruit 0</td>"`;

exports[`Component Table > should render a basic table > cells with border 2 1`] = `"<td style=\\"width: 150px;\\" class=\\"text-[13px] font-medium leading-[18px] text-text-text-and-icon px-4 truncate min-w-0 border-b border-bg-grid-border h-[48px]\\">Description for 0</td>"`;

exports[`Component Table > should render a basic table > cells without border 3 1`] = `"<td style=\\"width: 150px;\\" class=\\"text-[13px] font-medium leading-[18px] text-text-text-and-icon px-4 truncate min-w-0 h-[48px]\\">1</td>"`;

exports[`Component Table > should render a basic table > cells without border 4 1`] = `"<td style=\\"width: 150px;\\" class=\\"text-[13px] font-medium leading-[18px] text-text-text-and-icon px-4 truncate min-w-0 h-[48px]\\">Fruit 1</td>"`;

exports[`Component Table > should render a basic table > cells without border 5 1`] = `"<td style=\\"width: 150px;\\" class=\\"text-[13px] font-medium leading-[18px] text-text-text-and-icon px-4 truncate min-w-0 h-[48px]\\">Description for 1</td>"`;

exports[`Component Table > should render a basic table > row with hover state on normal table 1`] = `
<tr
  class="dark:hover:bg-bg-hover-2 hover:bg-bg-breadcrumb-bar transition-colors"
>
  <td
    class="text-[13px] font-medium leading-[18px] text-text-text-and-icon px-4 truncate min-w-0 h-[48px]"
    style="width: 150px;"
  >
    1
  </td>
  <td
    class="text-[13px] font-medium leading-[18px] text-text-text-and-icon px-4 truncate min-w-0 h-[48px]"
    style="width: 150px;"
  >
    Fruit 1
  </td>
  <td
    class="text-[13px] font-medium leading-[18px] text-text-text-and-icon px-4 truncate min-w-0 h-[48px]"
    style="width: 150px;"
  >
    Description for 1
  </td>
</tr>
`;

exports[`Component Table > should use getTrProps and getTdProps correctly > cells with custom props 0 1`] = `"<td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 border-b border-bg-grid-border h-[48px] text-red-500\\" style=\\"width: 150px;\\">0</td>"`;

exports[`Component Table > should use getTrProps and getTdProps correctly > cells with custom props 1 1`] = `"<td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 border-b border-bg-grid-border h-[48px] text-red-500\\" style=\\"width: 150px;\\">Fruit 0</td>"`;

exports[`Component Table > should use getTrProps and getTdProps correctly > cells with custom props 2 1`] = `"<td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 border-b border-bg-grid-border h-[48px] text-red-500\\" style=\\"width: 150px;\\">Description for 0</td>"`;

exports[`Component Table > should use getTrProps and getTdProps correctly > cells with custom props 3 1`] = `"<td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 h-[48px] text-red-500\\" style=\\"width: 150px;\\">1</td>"`;

exports[`Component Table > should use getTrProps and getTdProps correctly > cells with custom props 4 1`] = `"<td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 h-[48px] text-red-500\\" style=\\"width: 150px;\\">Fruit 1</td>"`;

exports[`Component Table > should use getTrProps and getTdProps correctly > cells with custom props 5 1`] = `"<td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 h-[48px] text-red-500\\" style=\\"width: 150px;\\">Description for 1</td>"`;

exports[`Component Table > should use getTrProps and getTdProps correctly > row with custom props0 1`] = `"<tr data-testid=\\"table-header-row\\"><th colspan=\\"1\\" class=\\"relative border-0 text-text-text-and-icon border-b-[1.5px] dark:border-bg-grid-border border-[#dcdcdc] text-p11\\" style=\\"width: 150px;\\"><div class=\\"w-full h-full flex truncate pl-4 pr-2.5 items-center py-[15px]\\"><span class=\\"flex-1 truncate text-start\\">id</span></div></th><th colspan=\\"1\\" class=\\"relative border-0 text-text-text-and-icon border-b-[1.5px] dark:border-bg-grid-border border-[#dcdcdc] text-p11\\" style=\\"width: 150px;\\"><div class=\\"w-full h-full flex truncate pl-4 pr-2.5 items-center py-[15px]\\"><span class=\\"flex-1 truncate text-start\\">name</span></div></th><th colspan=\\"1\\" class=\\"relative border-0 text-text-text-and-icon border-b-[1.5px] dark:border-bg-grid-border border-[#dcdcdc] text-p11\\" style=\\"width: 150px;\\"><div class=\\"w-full h-full flex truncate pl-4 pr-2.5 items-center py-[15px]\\"><span class=\\"flex-1 truncate text-start\\">description</span></div></th></tr>"`;

exports[`Component Table > should use getTrProps and getTdProps correctly > row with custom props1 1`] = `"<tr class=\\"dark:hover:bg-bg-hover-2 hover:bg-bg-breadcrumb-bar transition-colors bg-red-500\\"><td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 border-b border-bg-grid-border h-[48px] text-red-500\\" style=\\"width: 150px;\\">0</td><td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 border-b border-bg-grid-border h-[48px] text-red-500\\" style=\\"width: 150px;\\">Fruit 0</td><td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 border-b border-bg-grid-border h-[48px] text-red-500\\" style=\\"width: 150px;\\">Description for 0</td></tr>"`;

exports[`Component Table > should use getTrProps and getTdProps correctly > row with custom props2 1`] = `"<tr class=\\"dark:hover:bg-bg-hover-2 hover:bg-bg-breadcrumb-bar transition-colors bg-red-500\\"><td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 h-[48px] text-red-500\\" style=\\"width: 150px;\\">1</td><td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 h-[48px] text-red-500\\" style=\\"width: 150px;\\">Fruit 1</td><td class=\\"text-[13px] font-medium leading-[18px] px-4 truncate min-w-0 h-[48px] text-red-500\\" style=\\"width: 150px;\\">Description for 1</td></tr>"`;
